The `forall` function triggers `-Wforall-identifier` on GHC HEAD

Building `constraints` with GHC HEAD triggers a `-Wforall-identifier` warning:

```
[3 of 7] Compiling Data.Constraint.Forall ( src/Data/Constraint/Forall.hs, /home/ryanglscott/Documents/Hacking/Haskell/ci-maintenance/checkout/dist-newstyle/build/x86_64-linux/ghc-9.3.20220316/constraints-0.14/build/Data/Constraint/Forall.o, /home/ryanglscott/Documents/Hacking/Haskell/ci-maintenance/checkout/dist-newstyle/build/x86_64-linux/ghc-9.3.20220316/constraints-0.14/build/Data/Constraint/Forall.dyn_o )

src/Data/Constraint/Forall.hs:55:1: warning: [-Wforall-identifier]
    The use of ‘forall’ as an identifier
    will become an error in a future GHC release.
    Suggested fix:
      Consider using another name, such as
      ‘forAll’, ‘for_all’, or ‘forall_’.
   |
55 | forall d = forallish (unsafeCoerce d)
   | ^^^^^^
```

We should rename `forall` to something else (e.g., `forall_`) to avoid this warning. Note that `forall` is part of the public `Data.Constraint.Forall` API, so this would entail a breaking change. Since the next `constraints` release will require a major version anyway, we should do this before the next release.

--- a/CHANGELOG.markdown
+++ b/CHANGELOG.markdown
@@ -1,6 +1,9 @@
 0.14 [unreleased]
 -----------------
 * `Data.Constraint.Forall` uses `QuantifiedConstraints` on GHC 8.6+
+* The `forall` function in `Data.Constraint.Forall` has been renamed to
+  `forall_`, since a future version of GHC will make the use of `forall` as
+  an identifier an error.
 * Add `c => Boring (Dict c)` instance
 * Remove `Lifting` instances for `ErrorT` and `ListT`, which were removed
   in `transformers-0.6.*`.

--- a/src/Data/Constraint/Forall.hs
+++ b/src/Data/Constraint/Forall.hs
@@ -32,7 +32,7 @@ module Data.Constraint.Forall
   , Forall1, inst1
   , ForallT, instT
   , ForallV, InstV (instV)
-  , forall
+  , forall_
   ) where
 
 import Data.Constraint
@@ -51,8 +51,8 @@ data Dict1 p where
 forallish :: forall p. Dict1 p -> Dict (Forall p)
 forallish Dict1 = Dict
 
-forall :: forall p. (forall a. Dict (p a)) -> Dict (Forall p)
-forall d = forallish (unsafeCoerce d)
+forall_ :: forall p. (forall a. Dict (p a)) -> Dict (Forall p)
+forall_ d = forallish (unsafeCoerce d)
 
 -- | Composition for constraints.
 class p (f a) => ComposeC (p :: k2 -> Constraint) (f :: k1 -> k2) (a :: k1)

